src/NATS.Client.Core/AsyncEventHandler.cs

```diff
@@ -0,0 +1,35 @@
+﻿// origin: https://github.com/microsoft/vs-threading/blob/9065e6e4b5593e6ed6e3ff0a9159d4e2765430d6/src/Microsoft.VisualStudio.Threading/AsyncEventHandler.cs
+// license: MIT
+//
+// Microsoft.VisualStudio.Threading
+// Copyright (c) Microsoft Corporation
+// All rights reserved.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+//     of this software and associated documentation files (the "Software"), to deal
+//     in the Software without restriction, including without limitation the rights
+//     to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+//     copies of the Software, and to permit persons to whom the Software is
+//     furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+//     copies or substantial portions of the Software.
+
+namespace NATS.Client.Core;
+
+/// <summary>
+/// An asynchronous event handler.
+/// </summary>
+/// <param name="sender">The sender of the event.</param>
+/// <param name="args">Event arguments.</param>
+/// <returns>A task whose completion signals handling is finished.</returns>
+public delegate Task AsyncEventHandler(object? sender, EventArgs args);
```

Comment:
I don't see a downside personally. We might as well start with `ValueTask` to avoid breaking it later.

Reply:
I think using ValueTask is a mistake. This code is not on a hot path (if we had a MessageDelivered event, this would be a different story), and the awaiting code is outside the library - you must never await a ValueTask more than once, and this provides a gotcha for consumers of the library.

I'm not opposed to making it ValueTask per se (and will happily adjust the PR), but don't see this as an area that needs any optimization. There's only 4 events and none of them are expected to fire very often; if they do, you have more serious concerns to deal with.

Reply:
Good point about not being on hot path. In general we do use ValueTask pretty much everywhere though, hot path or not. If you don't see any technical problems using it (apart from double await) do you think we should consider using it for consistency as well?

Reply:
I would use ValueTask sparingly, unless it was internally or on a hot path. Stephen Toub has a great [guide](https://devblogs.microsoft.com/dotnet/understanding-the-whys-whats-and-whens-of-valuetask/) on the topic.

Reply:
I wanted to revive this thread - I just scanned our public API and I think we use `ValueTask` everywhere.  It might be best to stick to that for consistency's sake.  External code wouldn't be awaiting events anyways, we control when the events run and are awaited.